plugins/modem-manager/fu-firehose-updater.c

```diff
@@ -76,6 +77,40 @@ fu_firehose_updater_log_message(const gchar *action, GBytes *msg)
 	g_debug("%s: %.*s", action, (gint)msg_size, msg_strsafe);
 }
 
+static GBytes *
+fu_firehose_read(FuFirehoseUpdater *self, guint timeout_ms, FuIOChannelFlags flags, GError **error)
+{
+	if (self->sahara != NULL) {
+		GByteArray *bytearr = NULL;
+		g_autoptr(GBytes) bytes = NULL;
+
+		bytearr = fu_sahara_loader_qdl_read(self->sahara, error);
+		if (bytearr == NULL)
+			return NULL;
+
+		bytes = g_bytes_new_take(bytearr->data, bytearr->len);
+		return g_steal_pointer(&bytes);
```

Comment:
i think the memory management is a bit funky here, i.e. I think you're leaking the GByteArray container. I think you want `return g_byte_array_free_to_bytes()` instead.

Reply:
thanks, fixed it